.github/actions/multi-approvers/src/multi-approvers.ts

```diff
@@ -0,0 +1,274 @@
+        team_slug: this.team,
+        username: login,
+      });
+      return (
+        ALLOWED_TEAM_MEMBER_ROLES.includes(response.data.role) &&
+        response.data.state === ACTIVE
+      );
+    } catch (err) {
+      if (err instanceof RequestError && err.status === 404) {
+        this.logDebug(
+          `Received 404 testing membership; assuming user is not a member: ${JSON.stringify(
+            err,
+          )}`,
+        );
+        // We can get here for a few known reasons:
+        // 1) The user is not a member
+        // 2) The team does not exist
+        // 3) Invalid token
+        // In all these cases, it's safe to return false.
+        return false;
+      }
+      throw err;
+    }
+  }
+
+  // Returns the number of approvals from members in the given list.
+  private async internalApprovedCount(
+    submittedReviews: PullRequestReview,
+    prLogin: string,
+  ): Promise<number> {
+    // Sort by chronological order.
+    const sortedReviews = submittedReviews.sort(
+      (a, b) =>
+        new Date(a.submitted_at || 0).getTime() -
+        new Date(b.submitted_at || 0).getTime(),
+    );
+    const reviewStateByLogin = new Map<string, string>();
+
+    for (const r of sortedReviews) {
+      const reviewerLogin = r.user!.login;
+
+      // Ignore the PR user.
+      if (reviewerLogin === prLogin) {
+        continue;
+      }
+
+      // Only consider internal users.
+      const isInternalUser = await this.isInternal(reviewerLogin);
+      if (!isInternalUser) {
+        continue;
+      }
+
+      // Set state if it does not exist.
+      if (!reviewStateByLogin.has(reviewerLogin)) {
+        reviewStateByLogin.set(reviewerLogin, r.state);
+        continue;
+      }
+
+      // Always update state if not approved.
+      if (reviewStateByLogin.get(reviewerLogin) !== APPROVED) {
+        reviewStateByLogin.set(reviewerLogin, r.state);
+        continue;
+      }
+
+      // Do not update approved state for a comment.
+      if (
+        reviewStateByLogin.get(reviewerLogin) === APPROVED &&
+        r.state !== COMMENTED
+      ) {
+        reviewStateByLogin.set(reviewerLogin, r.state);
+        continue;
+      }
+    }
+
+    return Array.from(reviewStateByLogin.values()).filter((s) => s === APPROVED)
+      .length;
+  }
+
+  /** Checks that approval requirements are satisfied. */
+  private async validateApprovers(): Promise<Result> {
+    const prResponse = await this.octokit.rest.pulls.get({
+      owner: this.repoOwner,
+      repo: this.repoName,
+      pull_number: this.pullNumber,
+    });
+    const prLogin = prResponse.data.user.login;
+
+    const isInternalPr = await this.isInternal(prLogin);
+    if (isInternalPr) {
+      // Do nothing if the pull request owner is an internal user.
+      this.logInfo(
+        `Pull request login ${
+          prLogin
+        } is an internal member, therefore no special approval rules apply.`,
+      );
+      return Success.INSTANCE;
```

Comment:
This is a very strange pattern to me. Why are we returning a custom `Result` type with these constants over just returning void and throwing errors on error? It seems like we're re-invented errors for some reason, and I don't understand it.

Reply:
I removed this pattern.

If you're curious, I initially had the code as you suggest. But I then tried to not drive the logic with exceptions. In this case, I agree that it's better to keep it exception-driven.